From d2a06c0f456261c4bee93231d90af2f675169bf1 Mon Sep 17 00:00:00 2001
From: Ryan McCabe <rmccabe@redhat.com>
Date: Wed, 4 Mar 2015 08:01:53 -0500
Subject: [PATCH] luci: Add support for rrp_problem_count_threshold

Add support for setting/viewing rrp_problem_count_threshold in
expert mode. It appears under the totem config tab.

Resolves: rhbz#917773

Signed-off-by: Ryan McCabe <rmccabe@redhat.com>
---
 luci/lib/ClusterConf/Totem.py            | 11 ++++++++++-
 luci/templates/configure.html            |  1 +
 luci/validation/validate_cluster_prop.py | 11 ++++++++++-
 3 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/luci/lib/ClusterConf/Totem.py b/luci/lib/ClusterConf/Totem.py
index f98b5f7..6d9f2a7 100644
--- a/luci/lib/ClusterConf/Totem.py
+++ b/luci/lib/ClusterConf/Totem.py
@@ -1,4 +1,4 @@
-# Copyright (C) 2006-2014 Red Hat, Inc.
+# Copyright (C) 2006-2015 Red Hat, Inc.
 #
 # This program is free software; you can redistribute
 # it and/or modify it under the terms of version 2 of the
@@ -144,6 +144,15 @@ class Totem(TagObject):
   def delSeqnoUnchangedConst(self):
     return self.removeAttribute('seqno_unchanged_const')
 
+  def getRRPProblemCountThreshold(self):
+    return self.getAttribute('rrp_problem_count_threshold')
+
+  def setRRPProblemCountThreshold(self, val):
+    return self.addIntegerAttribute('rrp_problem_count_threshold', val, (0, None))
+
+  def delRRPProblemCountThreshold(self):
+    return self.removeAttribute('rrp_problem_count_threshold')
+
   def removeDefaults(self):
     if self.getJoinTimeout() == self.DEFAULTS.get('join'):
         self.delJoinTimeout()
diff --git a/luci/templates/configure.html b/luci/templates/configure.html
index 086b73d..6071fcd 100644
--- a/luci/templates/configure.html
+++ b/luci/templates/configure.html
@@ -181,6 +181,7 @@
                 <div class="row"><label class="wide">Network MTU</label><input name="netmtu" class="text" type="text" py:attrs="totem and { 'value': totem.getNetMTU() } or {}"/></div>
                 <div class="row"><label class="wide">Window Size</label><input name="window_size" class="text" type="text" py:attrs="totem and { 'value': totem.getWindowSize() } or {}"/></div>
                 <div class="row"><label class="wide">Sequence Number Unchanged Constant</label><input name="seqno_unchanged_const" class="text" type="text" py:attrs="totem and { 'value': totem.getSeqnoUnchangedConst() } or {}"/></div>
+                <div class="row"><label class="wide">RRP Problem Count Threshold</label><input name="rrp_problem_count_threshold" class="text" type="text" py:attrs="totem and { 'value': totem.getRRPProblemCountThreshold() } or {}"/></div>
                 <div class="row"><label class="wide">Encrypt Totem Network Traffic</label><input name="secauth" type="checkbox" class="checkbox" py:attrs="(not totem or totem.getSecAuth()) and {'checked':'checked'} or {}"/></div>
             </div>
             <div class="row"><input type="submit" class="button formsubmit blue" value="Apply"/>
diff --git a/luci/validation/validate_cluster_prop.py b/luci/validation/validate_cluster_prop.py
index dda8045..a5d0ec5 100644
--- a/luci/validation/validate_cluster_prop.py
+++ b/luci/validation/validate_cluster_prop.py
@@ -1,4 +1,4 @@
-# Copyright (C) 2009-2014 Red Hat, Inc.
+# Copyright (C) 2009-2015 Red Hat, Inc.
 #
 # This program is free software; you can redistribute
 # it and/or modify it under the terms of version 2 of the
@@ -660,6 +660,15 @@ def validate_network_config(model, **kw):
     else:
         totem.delSeqnoUnchangedConst()
 
+    rrp_problem_count_threshold = kw.get('rrp_problem_count_threshold')
+    if rrp_problem_count_threshold and not rrp_problem_count_threshold.isspace():
+        try:
+            totem.setRRPProblemCountThreshold(rrp_problem_count_threshold)
+        except:
+            errors.append(_('Invalid value for rrp problem count threshold: %s') % rrp_problem_count_threshold)
+    else:
+        totem.delRRPProblemCountThreshold()
+
     totem.setSecAuth(kw.get('secauth') is not None)
     return (len(errors) < 1, {'errors': errors})
 
-- 
2.1.0

